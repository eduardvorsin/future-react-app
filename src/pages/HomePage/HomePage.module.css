.home-page{
	position: relative;
}

.home-page__hero{
	/* .home-page__hero styles */
}

.hero-section{
	display: flex;
	align-items: center;
	padding: 30px 0;
	background-image: url(../../assets/images/books-bg.webp);
	background-color: var(--dark-color-100-30);
	background-position: 0 0;
	background-size: cover;
	background-repeat: no-repeat;
	background-blend-mode: color;
	width: 100%;
	min-height: 400px;
	overflow: hidden;
}

.hero-section__container{
	/* .hero-section__container styles */
}

.hero-section__title{
	margin-bottom: 25px;
}

.hero-section__content{
	max-width: 450px;
	width: 100%;
	margin:  0 auto;
}

.hero-section__search{
	margin-bottom: 15px;
	max-width: 100%;
}

.hero-section__category-select{
	margin-bottom: 15px;
}

.hero-section__sort-select{
	/* .hero-section__sort-select styles */
}

.books-section{
	padding: 30px 0;
	position: relative;
}

.books-section__error{
	margin: 0 auto;
}

.books-section__container{
	display: flex;
	flex-direction: column;
}

.books-section__button{
	align-self: center;
	margin-top: 20px;
}

@media (min-width:576px){
	.books-section{
		padding: 40px 0;
	}

	.hero-section{
		padding: 40px 0;
	}

	.hero-section__title{
		margin-bottom: 40px;
	}

	.hero-section__search{
		margin-bottom: 20px;
	}

	.hero-section__category-select{
		margin-bottom: 20px;
	}
}
