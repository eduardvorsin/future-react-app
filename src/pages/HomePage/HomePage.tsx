import React, { useState } from 'react';
import { useNavigate, useOutlet } from 'react-router-dom';
import useAppDispatch from '../../hooks/useAppDispatch/useAppDispatch';
import { BookCategories, BookSortOrder } from '../../store/bookSlice/bookSlice';
import fetchBooks from '../../store/thunks/fetchBooks/fetchBooks';
import SearchBar from '../../components/UI/SearchBar/SearchBar';
import Select from '../../components/UI/Select/Select';
import Title from '../../components/UI/Title/Title';
import classes from './HomePage.module.css';
import useAppSelector from '../../hooks/useAppSelector/useAppSelector';
import BooksCount from '../../components/BL/BooksCount/BooksCount';
import BooksList from '../../components/BL/BooksList/BooksList';
import Spinner from '../../components/UI/Spinner/Spinner';
import Error from '../../components/UI/Error/Error';
import Button from '../../components/UI/Button/Button';

type CategoryOption = {
  value: BookCategories,
  label: string
}

type SorterOrderOption = {
  value: BookSortOrder,
  label: string
}

const categoriesOptions: CategoryOption[] = [
  {
    value: 'all',
    label: 'Все',
  },
  {
    value: 'art',
    label: 'Исскуство',
  },
  {
    value: 'biography',
    label: 'Биография',
  },
  {
    value: 'computers',
    label: 'Компьютеры',
  },
  {
    value: 'history',
    label: 'История',
  },
  {
    value: 'medical',
    label: 'Медицина',
  },
  {
    value: 'poetry',
    label: 'Поэзия',
  },
];

const sortOrderOptions: SorterOrderOption[] = [
  {
    value: 'relevance',
    label: 'Релевантности',
  },
  {
    value: 'newest',
    label: 'Новейшим',
  },
];

const HomePage = () => {
  const [searchValue, setSearchValue] = useState<string>('');
  const [category, setCategory] = useState<string>(categoriesOptions[0].value);
  const [sortOrder, setSortOrder] = useState<string>(sortOrderOptions[0].value);

  const dispatch = useAppDispatch();
  const outlet = useOutlet();
  const navigate = useNavigate();

  const { data, status, error } = useAppSelector((state) => state.books);

  const changeHandler: React.ChangeEventHandler<HTMLInputElement> = (e) => {
    setSearchValue(e.target.value);
  };

  const searchHandler: () => void = () => {
    dispatch(fetchBooks({
      page: 0,
      name: searchValue,
      sortOrder,
      category,
    }));
    navigate('/');
  };

  const clearHandler: () => void = () => {
    setSearchValue('');
  };

  const sortOrderChangeHandler: React.ChangeEventHandler<HTMLSelectElement> = (e) => {
    setSortOrder(e.currentTarget.value);
  };

  const categoryChangeHandler: React.ChangeEventHandler<HTMLSelectElement> = (e) => {
    setCategory(e.currentTarget.value);
  };

  const loadMoreBooks = () => {
    dispatch(fetchBooks({
      page: data.page + 1,
      name: searchValue,
      sortOrder,
      category,
    }));
    navigate('/');
  };

  const currentError = error || 'Не удалось загрузить книги попробуйте еще раз';

  return (
    <div
      className={`${classes['home-page']}`}
    >
      <section
        className={`${classes['home-page__hero']} ${classes['hero-section']}`}
      >
        <div
          className={`${classes['hero-section__container']} container`}

        >
          <Title
            className={classes['hero-section__title']}
            variant='light'
            level={2}
          >
            Поиск книг
          </Title>

          <div
            className={classes['hero-section__content']}
          >
            <SearchBar
              className={classes['hero-section__search']}
              value={searchValue}
              onClear={clearHandler}
              onSearch={searchHandler}
              onChange={changeHandler}
            />

            <Select
              className={classes['hero-section__category-select']}
              value={category}
              labelText='Категории'
              id='categories'
              onChange={categoryChangeHandler}
              options={categoriesOptions}
            />

            <Select
              className={classes['hero-section__sort-select']}
              value={sortOrder}
              labelText='Отсортировать по'
              id='sortingBy'
              onChange={sortOrderChangeHandler}
              options={sortOrderOptions}
            />
          </div>
        </div>
      </section>

      <section
        className={classes['books-section']}
      >
        <div
          className={`${classes['books-section__container']} container`}
        >
          {status === 'loading' && (
            <Spinner
              size={200}
            />
          )}

          {status === 'rejected' && (
            <Error
              className={classes['books-section__error']}
              message={currentError}
            />
          )}

          {status === 'resolved' && outlet === null && (
            <>
              <BooksCount>Найдено {data.totalItems} книг</BooksCount>
              <BooksList
                data={data.items}
              />
              {data.totalItems - data.items.length > 0 && (
                <Button
                  className={classes['books-section__button']}
                  onClick={loadMoreBooks}
                >
                  Смотреть еще
                </Button>
              )}
            </>
          )}
          {outlet}
        </div>
      </section>
    </div>
  );
};

export default HomePage;
