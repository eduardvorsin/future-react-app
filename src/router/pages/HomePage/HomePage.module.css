.home-page {
	position: relative;
}

.home-page__hero {
	/* .home-page__hero styles */
}

.home-page__quick-menu {
	position: absolute;
	right: 5px;
	top: 5px;
	background-color: var(--secondary-color-100);
	max-width: 190px;
	width: auto;
	padding: 5px;
	display: flex;
	align-items: center;
}

.home-page__quick-menu-item:not(:last-child) {
	margin-right: 15px;
	margin-bottom: 0;
}

.hero-section {
	display: flex;
	align-items: center;
	padding: 75px 0 30px;
	background-image: url(../../../assets/images/books-bg.webp);
	background-color: var(--overlay-color-100-40);
	background-position: 0 0;
	background-size: cover;
	background-repeat: no-repeat;
	background-blend-mode: color;
	width: 100%;
	min-height: 400px;
	overflow: hidden;
}

.hero-section__container {
	flex-grow: 1;
}

.hero-section__title {
	color: var(--light-color-200);
	margin-bottom: 25px;
}

.hero-section__content {
	max-width: 350px;
	width: 100%;
	margin: 0 auto;
}

.books-section {
	padding: 30px 0;
	position: relative;
}

.books-section__error {
	margin: 0 auto;
}

.books-section__container {
	display: flex;
	flex-direction: column;
}

.books-section__button.books-section__button {
	align-self: center;
	margin-top: 30px;
}

@media (min-width:576px) {
	.books-section {
		padding: 40px 0;
	}

	.hero-section {
		padding: 75px 0 40px;
	}

	.hero-section__content {
		max-width: 530px;
	}

	.hero-section__title {
		margin-bottom: 40px;
	}

	.hero-section__search {
		margin-bottom: 20px;
	}

	.hero-section__category-select {
		margin-bottom: 20px;
	}
}

@media (min-width:992px) {
	.hero-section {
		padding: 135px 0 40px;
	}

	.home-page__quick-menu {
		padding: 10px;
		max-width: 130px;
		flex-direction: column;
	}

	.home-page__quick-menu-item:not(:last-child) {
		margin-bottom: 15px;
		margin-right: 0;
	}
}