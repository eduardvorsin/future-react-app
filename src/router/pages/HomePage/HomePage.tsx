import { useTranslation } from 'react-i18next';
import useAppDispatch from '../../../hooks/useAppDispatch/useAppDispatch';
import fetchBooks from '../../../store/thunks/fetchBooks/fetchBooks';
import Title from '../../../components/UI/Title/Title';
import classes from './HomePage.module.css';
import useAppSelector from '../../../hooks/useAppSelector/useAppSelector';
import BooksCount from '../../../components/BL/BooksCount/BooksCount';
import BooksList from '../../../components/BL/BooksList/BooksList';
import Spinner from '../../../components/UI/Spinner/Spinner';
import Error from '../../../components/UI/Error/Error';
import Button from '../../../components/UI/Button/Button';
import BooksSearch from '../../../components/BL/BooksSearch/BooksSearch';
import { setSearchPage } from '../../../store/bookSlice/bookSlice';
import { getVerticalScrollPosition, saveVerticalScrollPosition } from '../../../helpers/helpers';
import ThemeSwitcher from '../../../components/UI/ThemeSwitcher/ThemeSwitcher';
import LanguageSelect from '../../../components/UI/LanguageSelect/LanguageSelect';

const HomePage = () => {
  const { t } = useTranslation();
  const dispatch = useAppDispatch();
  const outlet = useOutlet();
  const navigate = useNavigate();

  const {
    data,
    totalItems,
    status,
    error,
    searchOptions,
  } = useAppSelector((state) => state.books);

  const loadMoreBooks = () => {
    const {
      page,
      bookName,
      sortOrder,
      category,
      searchBy,
    } = searchOptions;

    dispatch(fetchBooks({
      page: page + 1,
      bookName,
      sortOrder,
      category,
      searchBy,
    }));

    setSearchPage(page + 1);
    saveVerticalScrollPosition();
    navigate('/books');
  };

  useEffect(() => {
    window.scrollTo(0, Number(getVerticalScrollPosition()));
  }, [data]);

  const currentError = error ?? t('fetchBooksError');

  return (
    <div
      className={`${classes['home-page']}`}
    >
      <div
        className={classes['home-page__quick-menu']}
      >
        <div
          className={classes['home-page__quick-menu-item']}
        >
          <ThemeSwitcher />
        </div>
        <div
          className={classes['home-page__quick-menu-item']}
        >
          <LanguageSelect
          />
        </div>
      </div>

      <section
        className={`${classes['home-page__hero']} ${classes['hero-section']}`}
      >
        <div
          className={`${classes['hero-section__container']} container`}

        >
          <Title
            className={classes['hero-section__title']}
            component='h2'
            level={2}
          >
            {t('mainTitle', { ns: 'homePage' })}
          </Title>

          <BooksSearch
            className={classes['hero-section__content']}
          />
        </div>
      </section>
      <section
        className={classes['books-section']}
      >
        <div
          className={`${classes['books-section__container']} container`}
        >
          {status === 'loading' && (
            <Spinner
              size={200}
            />
          )}

          {status === 'rejected' && (
            <Error
              className={classes['books-section__error']}
              message={currentError}
            />
          )}

          {status === 'resolved' && outlet === null && (
            <>
              <BooksCount>
                {t('booksCount', {
                  count: totalItems,
                  currentBooksCount: data.length,
                  ns: 'homePage',
                })}
              </BooksCount>
              <BooksList
                data={data}
              />
              {totalItems - data.length > 0 && (
                <Button
                  className={classes['books-section__button']}
                  onClick={loadMoreBooks}
                >
                  {t('showMore')}
                </Button>
              )}
            </>
          )}

          {status === 'resolved' && outlet && (
            <>
              {outlet}
            </>
          )}

        </div>
      </section>
    </div>
  );
};

export default HomePage;
