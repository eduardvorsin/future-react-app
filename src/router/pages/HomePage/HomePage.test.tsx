import React from 'react';
import { render, screen } from '@testing-library/react';
import { RouterProvider, createMemoryRouter } from 'react-router-dom';
import HomePage from './HomePage';
import i18n from '../../../localization/i18next';
import ReduxWrapper from '../../../tests/helpers/ReduxWrapper';

const router = createMemoryRouter([{
  path: '/',
  element: <HomePage testId='test-home-page' />,
}]);

describe('HomePage tests', () => {
  it('is rendered correctly', () => {
    render(<RouterProvider router={router} />, { wrapper: ReduxWrapper });

    expect(screen.getByTestId<HTMLDivElement>('test-home-page')).toBeInTheDocument();
  });

  it('is a basic snapshot', () => {
    render(<RouterProvider router={router} />, { wrapper: ReduxWrapper });

    expect(screen.getByTestId<HTMLDivElement>('test-home-page')).toMatchSnapshot();
  });
});

describe('HomePage integration tests', () => {
  it('the component with the text in English is displayed correctly', () => {
    const localizedTitle = i18n.t('mainTitle', { ns: 'homePage' });

    render(<RouterProvider router={router} />, { wrapper: ReduxWrapper });

    expect(screen.getByText<HTMLHeadingElement>(localizedTitle)).toBeInTheDocument();
  });

  it('the component with the text in Russian is displayed correctly', () => {
    i18n.changeLanguage('ru');
    const localizedTitle = i18n.t('mainTitle', { ns: 'homePage' });

    render(<RouterProvider router={router} />, { wrapper: ReduxWrapper });

    expect(screen.getByText<HTMLHeadingElement>(localizedTitle)).toBeInTheDocument();
  });
});
