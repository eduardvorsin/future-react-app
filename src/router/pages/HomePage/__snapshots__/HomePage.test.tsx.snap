// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`HomePage tests is a basic snapshot 1`] = `
<div
  class="home-page"
  data-testid="test-home-page"
>
  <div
    class="home-page__quick-menu"
  >
    <div
      class="home-page__quick-menu-item"
    >
      <fieldset
        class="switcher theme-switcher undefined"
      >
        <legend
          class="switcher__label sr-only"
        >
          Color theme switcher
        </legend>
        <label
          class="switcher__input-label sr-only"
          for=":r1:-input-first"
        >
          светлая тема
        </label>
        <input
          aria-checked="false"
          class="switcher__radio"
          id=":r1:-input-first"
          name="theme-switcher"
          role="switch"
          type="radio"
          value="light"
        />
        <label
          class="switcher__input-label sr-only"
          for=":r1:-input-second"
        >
          темная тема
        </label>
        <input
          aria-checked="true"
          checked=""
          class="switcher__radio"
          id=":r1:-input-second"
          name="theme-switcher"
          role="switch"
          type="radio"
          value="dark"
        />
        <span
          class="switcher__status"
        />
      </fieldset>
    </div>
    <div
      class="home-page__quick-menu-item"
    >
      <div
        class="language__select undefined select"
      >
        <label
          class="select__label"
          for="language-select"
        >
          Language selection
        </label>
        <div
          class="select-wrapper"
        >
          <select
            class="select-wrapper__select"
            id="language-select"
          >
            <option
              value="en"
            >
              ENG
            </option>
            <option
              value="ru"
            >
              RUS
            </option>
          </select>
        </div>
      </div>
    </div>
  </div>
  <section
    class="home-page__hero hero-section"
  >
    <div
      class="hero-section__container container"
    >
      <h2
        class="title hero-section__title title--dark title--level-2"
      >
        Book Search
      </h2>
      <div
        class="books-search hero-section__content"
      >
        <form
          class="books-search__search-bar search-form search-form__search-bar"
        >
          <div
            class="search-bar"
            role="search"
          >
            <label
              class="search-bar__label sr-only"
              for="search-bar"
            >
              Search Bar
            </label>
            <input
              class="search-bar__input"
              id="search-bar"
              placeholder="Find a book..."
              value=""
            />
            <button
              class="button--icon-button search-bar__clear"
            >
              <svg />
              Clear
            </button>
            <button
              class="button--icon-button search-bar__find"
              type="submit"
            >
              <svg />
              To find
            </button>
          </div>
        </form>
        <div
          class="books-search__category-select books-search__select select"
        >
          <label
            class="select__label"
            for="categories"
          >
            Categories
          </label>
          <div
            class="select-wrapper"
          >
            <select
              class="select-wrapper__select"
              id="categories"
            >
              <option
                value="all"
              >
                All
              </option>
              <option
                value="art"
              >
                Art
              </option>
              <option
                value="biography"
              >
                Biography
              </option>
              <option
                value="computers"
              >
                Computers
              </option>
              <option
                value="history"
              >
                History
              </option>
              <option
                value="medical"
              >
                Medical
              </option>
              <option
                value="poetry"
              >
                Poetry
              </option>
            </select>
          </div>
        </div>
        <div
          class="books-search__sort-select books-search__select select"
        >
          <label
            class="select__label"
            for="sortingBy"
          >
            Sort by
          </label>
          <div
            class="select-wrapper"
          >
            <select
              class="select-wrapper__select"
              id="sortingBy"
            >
              <option
                value="relevance"
              >
                Relevance
              </option>
              <option
                value="newest"
              >
                Newest
              </option>
            </select>
          </div>
        </div>
        <div
          class="books-search__find-by-select books-search__select select"
        >
          <label
            class="select__label"
            for="findBy"
          >
            Search by
          </label>
          <div
            class="select-wrapper"
          >
            <select
              class="select-wrapper__select"
              id="findBy"
            >
              <option
                value="intitle"
              >
                By title
              </option>
              <option
                value="inauthor"
              >
                By author
              </option>
              <option
                value="inpublisher"
              >
                By publisher
              </option>
              <option
                value="isbn"
              >
                By ISBN
              </option>
            </select>
          </div>
        </div>
      </div>
    </div>
  </section>
  <section
    class="books-section"
  >
    <div
      class="books-section__container container "
    />
  </section>
</div>
`;
